rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -0,0 +1,590 @@
+        }
+
+        let mut counting_buf = CountBuf::new(buffer);
+        let name_len = counting_buf.get_u8();
+        if name_len as usize >= counting_buf.remaining() {
+            return Err(Error::InvalidHeaderNameLength);
+        }
+
+        let name = counting_buf.copy_to_bytes(name_len as usize);
+        let value = HeaderValue::read_from(&mut counting_buf)?;
+        Ok((Header::new(name, value), counting_buf.into_count()))
+    }
+
+    /// Writes the header to the given `buffer`.
+    fn write_to(&self, buffer: &mut dyn BufMut) -> Result<(), Error> {
+        if self.name.as_bytes().len() > MAX_HEADER_NAME_LEN {
+            return Err(Error::InvalidHeaderNameLength);
+        }
+
+        buffer.put_u8(u8::try_from(self.name.as_bytes().len()).expect("bounds check above"));
+        buffer.put_slice(&self.name.as_bytes()[..]);
+        self.value.write_to(buffer)
+    }
+}
+
+/// Event Stream message.
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq)]
+pub struct Message {
+    headers: Vec<Header>,
+    payload: Bytes,
+}
+
+impl Message {
+    /// Creates a new message with the given `payload`. Headers can be added later.
+    pub fn new(payload: impl Into<Bytes>) -> Message {
+        Message {
+            headers: Vec::new(),
+            payload: payload.into(),
+        }
+    }
+
+    /// Adds a header to the message.
+    pub fn add_header(mut self, header: Header) -> Self {
+        self.headers.push(header);
+        self
+    }
+
+    /// Returns all headers.
+    pub fn headers(&self) -> &[Header] {
+        &self.headers
+    }
+
+    /// Returns the payload bytes.
+    pub fn payload(&self) -> &Bytes {
+        &self.payload
+    }
+
+    // Returns (total_len, header_len)
+    fn read_prelude_from(buffer: &mut dyn Buf) -> Result<Option<(u32, u32)>, Error> {
+        let mut crc_buffer = CrcBuf::new(buffer);
+
+        // If the buffer doesn't have the entire frame yet, then short circuit
+        let total_len = crc_buffer.get_u32();
+        if crc_buffer.remaining() + size_of::<u32>() < total_len as usize {
+            return Ok(None);
+        }
+
+        // Validate the prelude
+        let header_len = crc_buffer.get_u32();
+        let (expected_crc, prelude_crc) = (crc_buffer.into_crc(), buffer.get_u32());
+        if expected_crc != prelude_crc {
+            return Err(Error::PreludeChecksumMismatch(expected_crc, prelude_crc));
+        }
+        // The header length can be 0 or >= 2, but must fit within the frame size
+        if header_len == 1 || header_len > max_header_len(total_len)? {
+            return Err(Error::InvalidHeadersLength);
+        }
+        Ok(Some((total_len, header_len)))
+    }
+
+    /// Reads a message from the given `buffer`. If the buffer doesn't have the whole
+    /// message in it yet (as is the case when streaming), then it returns `Ok(None)`.
+    pub fn read_from(buffer: &mut dyn Buf) -> Result<Option<Message>, Error> {
```

Comment:
I have a slight preference for explicit enums instead of `Option` here:
```rust
enum ParsedMessage {
  Complete(Message),
  Incomplete
}
```

naming bikeshedding welcome

Reply:
we can follow tokio, although Tokio also needs special docs to clarify what it means...since None can imply end of stream OR partial message it might be worth a distinction